Do not shadow variables (gcc set)

paveljanik opened the pull request:
Followup to #8105.

The current set of PRs made `clang` compile `Wshadow` clean. Compiling with `gcc` revealed new needed changes. `gcc` is more strict in warnings. This is non-Qt part. I do not have an environment to test `gcc`+Qt build now.

paveljanik committed:
Check and enable -Wshadow by default.

paveljanik commented:
Hmm, two failures in `p2p-segwit.py`. Is there some issue with this test script?

MarcoFalke commented:
Travis fails reliably. Is the commit producing identical binaries?

paveljanik commented:
This commit should produce same binaries, yes.

paveljanik commented on src/chain.h:
```diff
@@ -357,9 +357,9 @@ class CDiskBlockIndex : public CBlockIndex
-    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int _nVersion) {
         if (!(nType & SER_GETHASH))
-            READWRITE(VARINT(nVersion));
```
I'm investigating this hunk...

paveljanik commented on src/primitives/block.h:
```diff
@@ -92,7 +92,7 @@ class CBlock : public CBlockHeader
 
     template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int _nVersion) {
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
```
Hmm, this particular change is causing my clang on OS X build to fail `-rescan` with:

```
2016-09-26 19:53:08 ERROR: ReadBlockFromDisk: Deserialize or I/O error - ReadCompactSize(): size too large: unspecified iostream_category error at CBlockDiskPos(nFile=49, nPos=4208853)
```

So reverting it.

But do not ask me, why it is the cause... I do not know yet.

laanwj replied:
I think I figured this out, see https://github.com/bitcoin/bitcoin/pull/8658#issuecomment-250096029 and the comment below that.

laanwj commented:
This introduces binary changes in these serialization functions:
```
void CDiskBlockIndex::SerializationOp<CSizeComputer, CSerActionSerialize>(CSizeComputer&, CSerActionSerialize, int, int)
void CDiskBlockIndex::SerializationOp<CDataStream, CSerActionSerialize>(CDataStream&, CSerActionSerialize, int, int)
void CMerkleTx::SerializationOp<CDataStream, CSerActionUnserialize>(CDataStream&, CSerActionUnserialize, int, int)
void CMerkleTx::SerializationOp<CDataStream, CSerActionSerialize>(CDataStream&, CSerActionSerialize, int, int)
void CWalletTx::SerializationOp<CDataStream, CSerActionSerialize>(CDataStream&, CSerActionSerialize, int, int)
void CDiskBlockIndex::SerializationOp<CDataStream, CSerActionUnserialize>(CDataStream&, CSerActionUnserialize, int, int)
void CWalletTx::SerializationOp<CDataStream, CSerActionUnserialize>(CDataStream&, CSerActionUnserialize, int, int)
```

paveljanik commented:
OK, rebased and removed problematic parts (all serialization related, `nVersion`)...

paveljanik commented:
Rebased.

paveljanik commented:
@MarcoFalke Marco, can you please test for the same binaries here?

MarcoFalke commented:
There seems to be a small diff:

```diff
319294c319294
<   17d0f1:	lea    0x2dd1b2(%rip),%rdi        # 45a2aa <leveldb::Status::Status(leveldb::Status::Code, leveldb::Slice const&, leveldb::Slice const&)::__PRETTY_FUNCTION__+0xaa>
---
>   17d0f1:	lea    0x2dd1b1(%rip),%rdi        # 45a2a9 <leveldb::Status::Status(leveldb::Status::Code, leveldb::Slice const&, leveldb::Slice const&)::__PRETTY_FUNCTION__+0xa9>
```

paveljanik commented:
This seems unrelated.

fanquake commented:
I haven't done the identical binary checking yet, but it looks like the difference could be [here](https://github.com/bitcoin/bitcoin/blob/master/src/leveldb/db/repair.cc#L169) ? 170 (0xaa) vs 169 (0xa9)

paveljanik commented:
@fanquake surely not...

paveljanik commented:
JFYI: I'm now building with other gcc version and there are a few other warnings 8)

gmaxwell commented on src/arith_uint256.cpp:
```diff
@@ -173,9 +173,9 @@ unsigned int base_uint<BITS>::bits() const
 {
     for (int pos = WIDTH - 1; pos >= 0; pos--) {
         if (pn[pos]) {
-            for (int bits = 31; bits > 0; bits--) {
```
change _bits to nbits.

paveljanik replied:
Will do.

gmaxwell commented on src/streams.h:
```diff
@@ -523,11 +523,11 @@ class CBufferedFile
             throw std::ios_base::failure(feof(src) ? "CBufferedFile::Fill: end of file" : "CBufferedFile::Fill: fread failed");
         } else {
-            nSrcPos += read;
+            nSrcPos += _read;
```
change _read to nBytes

MarcoFalke commented:
Concept ACK when this is the last patch/pull with regard to nShadow.

gmaxwell commented on src/timedata.h:
```diff
@@ -48,14 +48,14 @@ class CMedianFilter
-        int size = vSorted.size();
-        assert(size > 0);
-        if (size & 1) // Odd number of elements
+        int _size = vSorted.size();
```
vecsize

dcousens replied:
`vSortedSize` ?

paveljanik replied:
I like vSortedSize, will do.

gmaxwell commented on src/primitives/transaction.h:
```diff
@@ -293,7 +293,7 @@ inline void SerializeTransaction(TxType& tx, Stream& s, Operation ser_action, in
 
     READWRITE(*const_cast<int32_t*>(&tx.nVersion));
-    unsigned char flags = 0;
+    unsigned char _flags = 0;
```
sflags  (this was less obvious to me, but my suggestion and @sipa thought it was okay)

dcousens replied:
`serializeFlags`?

gmaxwell commented on src/wallet/wallet.cpp:
```diff
@@ -280,17 +280,17 @@ bool CWallet::LoadWatchOnly(const CScript &dest)
 {
     CCrypter crypter;
-    CKeyingMaterial vMasterKey;
+    CKeyingMaterial _vMasterKey;
```
So I believe changing the name of CCryptoKeyStore  master key will, I think change only four lines. Might be good to get an opinion from someone who's worked more on this code?  @TheBlueMatt @pstratem

paveljanik replied:
That would be great and even better solution, yes. Will wait.

gmaxwell reviewed (commented):
filling up locals with _ prefixes in big pieces of code is less good than using better names, and also incompatible with a common convention of using _prefixes for function parameters.

gmaxwell commented:
I checked each of the changes so far, they appear correct to me (+/- the suggestion I made about changing the name of the member vMasterKey that I asked for others input on).  I didn't check that they resulted in the same code.

paveljanik committed:
Do not shadow variables (gcc set)

paveljanik commented:
Rebased, updated.

It is now `-Wshadow` clean for both old gcc (tested on gcc version 4.8.3 20140627 [gcc-4_8-branch revision 212064] (SUSE Linux)) and also on the ultimate gcc (gcc version 6.2.1 20160830 [gcc-6-branch revision 239856] (SUSE Linux)).

MarcoFalke commented:
@paveljanik Mind to  cherry-pick fd5654c, so we don't have to open another pull?

MarcoFalke commented:
I get same binaries for bedf9a7 besides this diff:

```sh
commit=bedf9a7 && git checkout bitcoin/master && git reset --hard HEAD && curl https://raw.githubusercontent.com/laanwj/bitcoin-maintainer-tools/6e4425587736144b067f67ad792d9ee904e74fd7/patches/stripbuildinfo.patch | patch -p 1 && make -j 2 && objdump -d -r -C --no-show-raw-insn src/bitcoind > /tmp/d_old && curl https://github.com/bitcoin/bitcoin/commit/"${commit}".diff | patch -p 1 && make -j 2 && objdump -d -r -C --no-show-raw-insn src/bitcoind > /tmp/d_new && diff /tmp/d_old /tmp/d_new
```

```diff
317267c317267
<   17aa0b:	lea    0x25704b(%rip),%rsi        # 3d1a5d <boost::recursive_mutex::lock()::__PRETTY_FUNCTION__+0xfd>
---
>   17aa0b:	lea    0x25705b(%rip),%rsi        # 3d1a6d <boost::recursive_mutex::lock()::__PRETTY_FUNCTION__+0x10d>
317684c317684
<   17b325:	lea    0x256737(%rip),%rsi        # 3d1a63 <boost::recursive_mutex::lock()::__PRETTY_FUNCTION__+0x103>
---
>   17b325:	lea    0x256747(%rip),%rsi        # 3d1a73 <boost::recursive_mutex::lock()::__PRETTY_FUNCTION__+0x113>
317757c317757
<   17b471:	lea    0x2dc552(%rip),%rdi        # 4579ca <leveldb::Status::Status(leveldb::Status::Code, leveldb::Slice const&, leveldb::Slice const&)::__PRETTY_FUNCTION__+0xaa>
---
>   17b471:	lea    0x2565e5(%rip),%rdi        # 3d1a5d <boost::recursive_mutex::lock()::__PRETTY_FUNCTION__+0xfd>

paveljanik commented:
@MarcoFalke I can, but Wladimir doesn't want it to be enabled by default...

paveljanik commented:
@MarcoFalke Agreed. We should merge this and then fix the remaining issues found by users who use other build environments/configurations.

jtimon commented:
Concept ACK

paveljanik commented:
Working on needed rebase...

MarcoFalke commented:
utACK ad1ae7ae2e90a29fcfaaddee2d5acd025de72cbe, 9de90bb gives same binaries for me besides some boost_pretty_functions.

jtimon commented:
utACK ad1ae7a

laanwj commented:
Checked that this, rebased on master, results in identical executables. ACK ad1ae7a
